flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TaskMetricGroup.java

```diff
@@ -144,15 +144,17 @@ public OperatorMetricGroup addOperator(OperatorID operatorID, String name) {
 			name = name.substring(0, METRICS_OPERATOR_NAME_MAX_LENGTH);
 		}
 		OperatorMetricGroup operator = new OperatorMetricGroup(this.registry, this, operatorID, name);
+		// unique OperatorIDs only exist in streaming, so we have to rely on the name for batch operators
+		final String key = operatorID + name;
 
 		synchronized (this) {
-			OperatorMetricGroup previous = operators.put(operatorID, operator);
+			OperatorMetricGroup previous = operators.put(key, operator);
```

Comment:
I suggest to reduce the code under `synchronized` using:
```
return operators.putIfAbsent(key, operator);
```
Also, I would rename `addOperator` function name to `getOrAddOperatorMetric`.

Reply:
I will rename the method but leave the code as is. It was intentionally written that way so that we only do a single lookup on the happy path. The default implementation of `putIfAbsent` is just syntactic sugar for separate get/put calls. While the HashMap _implementation_ of this method is indeed more efficient in this regard this is an implementation detail.